### PR TITLE
[MI-2294][server]: Created plugin API to fetch possible values for subscription filters.

### PR DESCRIPTION
**Pull request by avas27JTG**
Created plugin API to fetch possible values for subscription filters

### CODE REVIEW

File: server/plugin/api.go [hunk 1]
```diff
@@ -787,6 +788,36 @@ func (p *Plugin) handleGetGitRepositoryBranches(w http.ResponseWriter, r *http.R
...
+
+	body, err := serializers.GetSubscriptionFilterPossibleValuesRequestPayloadFromJSON(r.Body)
+	if err != nil {
+		p.API.LogError("Error in decoding the body for fetching subscription filter possible values", "Error", err.Error())
+		p.handleError(w, r, &serializers.Error{Code: http.StatusBadRequest, Message: err.Error()})
+		return
+	}
+
+	if validationErr := body.IsSubscriptionRequestPayloadValid(); validationErr != nil {
+		p.handleError(w, r, &serializers.Error{Code: http.StatusBadRequest, Message: validationErr.Error()})
```
Comment (ayusht2810) [inline server/plugin/api.go:802]: don't we need to log an error here?

File: server/plugin/client.go [hunk 2]
```diff
@@ -236,6 +237,37 @@ func (c *client) GetGitRepositoryBranches(organization, projectName, repository,
...
+	var subscriptionFilters []serializers.SubscriptionFilter
+	for _, filter := range request.Filters {
+		subscriptionFilters = append(subscriptionFilters, serializers.SubscriptionFilter{InputID: filter})
+	}
+
+	subscriptionFiltersRequest := &serializers.GetSubscriptionFilterValuesRequestPayloadFromClient{
+		Subscription: &serializers.CreateSubscriptionBodyPayload{
+			PublisherID:      "tfs",
+			ConsumerID:       "webHooks",
+			ConsumerActionID: "httpRequest",
```
Comment (ayusht2810) [inline server/plugin/client.go:252]: i guess there are already constants present for these. please check once

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/api.go [hunk 3]
```diff
@@ -787,6 +788,36 @@ func (p *Plugin) handleGetGitRepositoryBranches(w http.ResponseWriter, r *http.R
...
+	}
+
+	subscriptionFilterValues, statusCode, err := p.Client.GetSubscriptionFilterPossibleValues(body, mattermostUserID)
+	if err != nil {
+		p.API.LogError(constants.ErrorFetchSubscriptionFilterPossibleValues, "Error", err.Error())
+		p.handleError(w, r, &serializers.Error{Code: statusCode, Message: err.Error()})
+		return
+	}
+
+	filterwiseResponse := make(map[string][]serializers.PossibleValues)
```
Comment (manojmalik20) [inline server/plugin/api.go:813]:
```suggestion
	filterwiseResponse := make(map[string][]*serializers.PossibleValues)
```

File: server/plugin/client.go [hunk 4]
```diff
@@ -236,6 +237,37 @@ func (c *client) GetGitRepositoryBranches(organization, projectName, repository,
 	return gitBranchesResponse, statusCode, nil
 }
 
+func (c *client) GetSubscriptionFilterPossibleValues(request *serializers.GetSubscriptionFilterPossibleValuesRequestPayload, mattermostUserID string) (*serializers.SubscriptionFilterPossibleValuesResponseFromClient, int, error) {
+	getSubscriptionFilterValuesURL := fmt.Sprintf(constants.GetSubscriptionFilterPossibleValues, request.Organization)
+
+	var subscriptionFilters []serializers.SubscriptionFilter
```
Comment (manojmalik20) [inline server/plugin/client.go:243]:
```suggestion
	var subscriptionFilters []*serializers.SubscriptionFilter
```

File: server/plugin/client.go [hunk 5]
```diff
@@ -236,6 +237,37 @@ func (c *client) GetGitRepositoryBranches(organization, projectName, repository,
...
+			PublisherID:      "tfs",
+			ConsumerID:       "webHooks",
+			ConsumerActionID: "httpRequest",
+			EventType:        request.EventType,
+			PublisherInputs: serializers.PublisherInputsGeneric{
+				ProjectID: request.ProjectID,
+			},
+		},
+		InputValues: subscriptionFilters,
+		Scope:       10,
```
Comment (manojmalik20) [inline server/plugin/client.go:259]: What's this scope here? Shouldn't this be a constant variable?

File: server/serializers/subscriptions.go [hunk 6]
```diff
@@ -14,6 +14,10 @@ type UserID struct {
 	UniqueName  string `json:"uniqueName"`
 }
 
+type PublisherInputsGeneric struct {
+	ProjectID string `json:"projectId"`
+}
+
 type PublisherInputsBoards struct {
 	ProjectID string `json:"projectId"`
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:22]:
```suggestion
	PublisherInputsGeneric
```
See if this works and if it does, do this for repos too.

File: server/serializers/subscriptions.go [hunk 7]
```diff
@@ -45,7 +49,6 @@ type SubscriptionList struct {
 	Count             int                 `json:"count"`
 	SubscriptionValue []SubscriptionValue `json:"value"`
 }
-
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:48]: Why this change? Revert it.

File: server/serializers/subscriptions.go [hunk 8]
```diff
@@ -57,6 +60,37 @@ type CreateSubscriptionRequestPayload struct {
...
+	Filters      []string `json:"filters"`
+}
+
+type SubscriptionFilter struct {
+	InputID string `json:"inputId"`
+}
+
+type GetSubscriptionFilterValuesRequestPayloadFromClient struct {
+	Subscription *CreateSubscriptionBodyPayload `json:"subscription"`
+	InputValues  []SubscriptionFilter           `json:"inputValues"`
```
Comment (manojmalik20) [inline server/serializers/subscriptions.go:76]:
```suggestion
	InputValues  []*SubscriptionFilter           `json:"inputValues"`
```
Wherever we are using array of struct, we should use pointers.

File: server/serializers/subscriptions.go (hunk 6 above)
Comment (avas27JTG) [reply server/serializers/subscriptions.go:22]: It didn't work due to a mixture of struct and fields
error message: `mixture of field:value and value elements in struct literal`

File: server/plugin/client.go (hunk 5 above)
Comment (avas27JTG) [reply server/plugin/client.go]: it is not going to be used anywhere else so, I think there is no need for a constant variable, I have added a comment for its role

**Review by ayusht2810 — APPROVED** [review]

**Review by manojmalik20 — APPROVED** [review]
